src/deno_dir.rs

```diff
@@ -453,19 +451,26 @@ fn test_code_cache() {
   let r = deno_dir.code_cache(filename, source_code, output_code, source_map);
   r.expect("code_cache error");
   assert!(cache_path.exists());
-  assert_eq!(output_code, fs::read_to_string(&cache_path).unwrap());
+  assert_eq!(output_code.as_ref(), &fs::read(&cache_path).unwrap()[..]);
 }
 
 // https://github.com/denoland/deno/blob/golang/deno_dir.go#L25-L30
-fn source_code_hash(filename: &str, source_code: &str) -> String {
+fn source_code_hash(filename: &str, source_code: &[u8]) -> String {
   let mut ctx = ring::digest::Context::new(&ring::digest::SHA1);
   ctx.update(filename.as_bytes());
-  ctx.update(source_code.as_bytes());
+  ctx.update(source_code);
   let digest = ctx.finish();
-  let mut out = String::new();
-  // TODO There must be a better way to do this...
+  let mut out = String::with_capacity(digest.as_ref().len() * 2);
+
+  const DIGIT_TO_CHAR: [char; 16] = [
+    '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e',
+    'f',
+  ];
   for byte in digest.as_ref() {
-    write!(&mut out, "{:02x}", byte).unwrap();
+    let low = (byte & 0x0F) as usize;
+    let high = (byte >> 4) as usize;
+    out.push(DIGIT_TO_CHAR[high]);
+    out.push(DIGIT_TO_CHAR[low]);
```

Comment:
I'm not sure this is better?

Reply:
It is worse in terms of readability. It is a tradeoff.

Reply:
What's the tradeoff? Is this faster?

Reply:
Yes, definitely faster. Formatter involves very complex execution path.

Reply:
I'm skeptical - certainly this is faster - but I doubt it's 2X faster. Anyway it doesn't matter. This is readable enough.

Reply:
I bet 10x faster. The default impl of `ToString` trait is based on formatter. It is quite slow. That is why many types has its own version of specialized impl. This was a big motivation of specilization feature.

Reply:
wow, interesting...

Reply:
I accidentally find that "Hex encoding" is something very suitable for SIMD, it is used as an example in [standard document](https://doc.rust-lang.org/std/arch/index.html#examples). My version is the `hex_encode_fallback`.

I would leave this optimization for future work.